design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,122 @@
+---
+title: certificates.k8s.io Adoption
+authors:
+  - "@joshvanl"
+reviewers:
+  - "@joshvanl"
+approvers:
+  - "@munnerz"
+  - @jetstack/team-cert-manager
+editor: "@joshvanl"
+creation-date: 2021-02-09
+last-updated: 2021-02-09
+status: provisional
+---
+
+# certificates.k8s.io Adoption
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Signers](#signers)
+  - [API Changes](#api-changes)
+  - [Upgrading](#upgrading)
+  - [Risks and Mitigations](#risks-and-mitigations)
+<!-- /toc -->
+
+## Summary
+
+In Kubernetes v1.19 the
+[`CertificateSigningRequest`](https://github.com/kubernetes/api/blob/48bd8381a38a486f8b3cdf28cf7334a45b182f2e/certificates/v1/types.go#L41)
+resource graduated to `certificates.k8s.io/v1`. This makes the concept of
+requesting, signing, and consuming certificates in Kubernetes first class
+objects. cert-manager is well placed to serve this resource type as it becomes
+more popular in the wider community, whilst preserving all the features and
+extensions that cert-manager has to offer.
+
+
+## Motivation
+
+With the extensiveness of cert-manager’s ecosystem, it is well placed to manage
+the core Kubernetes resource type. This gives cert-manager the ability to
+integrate with components that it otherwise wouldn't have, without changes to the
+third-party project.
+
+
+### Goals
+
+- Add `CertificateSigningRequest` signer controllers to cert-manager for all
```

Comment:
I'm not sure if I missed a discussion, but it seems unfortunate that existing external issuers will also have to add support for this. Would it make sense to read the contents of `certificates.k8s.io.CertificateSigningRequest` and create a matching `cert-manager.io/CertificateRequest` in cert-manager's cluster-resource namespace?

Reply:
That may be something to consider as a migration piece to make things easier. That said, given we extract user identity out of requests on k8s.io CSR objects (and are looking to do the same for CertificateRequest) this may be a little tough to do properly.

Additionally, the non-namespaced nature of CSR objects vs the namespaced CR make this especially difficult. I'd err on the side of caution (at least initially) and _not_ implement this piece for now, until we get some feedback on the difficulties people face. There's far fewer issuer implementations than there are users of the project :)